### PR TITLE
Implemented mmap

### PR DESCRIPTION
**Pull request by rafalcieslak**
This is a very basic `mmap` implementation that only supports anonymous mappings. Generally, it is merely a fancy wrapping for `vm_map_findspace` and `vm_map_add_entry`. It is not yet hooked to a syscall handler, but eventually it will be useful for userspace programs.

Incidentally I discovered some minor bugs in `vm_map_findspace`, so this branch fixes them as well.

### CODE REVIEW

**Comment by cahirwpz** [comment]
As far as I can tell `malloc` from `newlib` can be configured to use `mmap`. But since there's not a single `mmap` user for the time being let's postpone this commit. However fixes to `vm_map_findspace` are of high value right now, so move them to another PR, which will get merged very soon.

**Comment by rafalcieslak** [comment]
I've cleaned up this branch and updated it to use new syscall interface.

**Comment by cahirwpz** [comment]
I'll not accept `mmap` and `munmap` syscall interface until kernel-side interface is well tested.

Having kernel threads gracefully crash on bad address access and `copyin` and `copyout` routines in place, it's time to provide thorough testing for `vm_map` interface. With that being done and added implementation of signals we can return to `mmap` and family implementation.

**Comment by rafalcieslak** [comment]
Do we really need signals before a simple `mmap`? I still believe `mmap` will be useful for `fork` - user programs typically use `mmap` to allocate the stack of a new thread when asking `fork` to share memory map between parent/child.

File: sys/sysent.c [hunk 1]
```diff
@@ -42,6 +43,7 @@ int sys_exit(thread_t *td, syscall_args_t *args) {
 
 /* clang-format hates long arrays. */
 sysent_t sysent[] = {
-  {sys_nosys}, {sys_exit},  {sys_open},  {sys_close}, {sys_read},  {sys_write},
-  {sys_lseek}, {sys_nosys}, {sys_nosys}, {sys_nosys}, {sys_fstat}, {sys_sbrk},
+  {sys_nosys}, {sys_exit},  {sys_open},  {sys_close}, {sys_read},
```
Comment (cahirwpz) [inline sys/sysent.c]: Please use `[SYS_MMAP] = {sys_mmap}` extended C syntax to avoid problems with entry renumbering.

File: sys/sys_mmap.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      vm_map_add_entry(vmap, addr, addr + length, VM_PROT_READ | VM_PROT_WRITE);
+    /* Assign a pager. */
+    entry->object = default_pager->pgr_alloc();
+
+  } /* Release vm_map rwlock. */
+
+  log("Created entry at %p, length: %zu", (void *)addr, length);
+
+  if (flags & MMAP_FLAG_ANONYMOUS) {
+    /* Calling bzero here defeats the purpose of paging on demand, but I do not
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: Well... we should assume it zeros the pages. However that seems to be broken, please fix [default_pager_fault](http://mimiker.ii.uni.wroc.pl/source/xref/mimiker/sys/vm_pager.c#16). Call to  `vm_map_protect` won't be necessary then.

File: sys/sys_mmap.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    log("mmap called by a thread with no user space.");
+    /* I have no idea what would be a good error code for this case. */
+    *error = ENOENT;
+    return MMAP_FAILED;
+  }
+
+  {
+    rw_scoped_enter(&vmap->rwlock, RW_WRITER);
+
+    if (addr >= vmap->pmap->end) {
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: I guess this and next check could be moved outside the scope.

File: sys/sys_mmap.c [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      log("Non-anonymous memory mappings are not yet implemented.");
+      *error = EINVAL;
+      return MMAP_FAILED;
+    }
+
+    length = roundup(length, PAGESIZE);
+
+    if (addr == 0x0) {
+      /* We will choose an appropriate place in vm map. */
+      if (vm_map_findspace_nolock(vmap, MMAP_LOW_ADDR, length, &addr) != 0) {
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: Lines 56-59 are same as 69-73. Shouldn't you consider `addr != 0` case first?

File: sys/sys_mmap.c [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+           space. */
+        if (vm_map_findspace_nolock(vmap, MMAP_LOW_ADDR, length, &addr) != 0) {
+          /* Still no memory found. */
+          *error = ENOMEM;
+          return MMAP_FAILED;
+        }
+      }
+    }
+
+    /* TODO: Check if the range of addresses we are about to map is not already
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: Umm... what... why?

File: user/include/sys/mman.h [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#include <stdint.h>
+
+/* Newlib does not provide mmap prototype, so we need to use our own. */
+
+#define MMAP_FLAG_ANONYMOUS 1
```
Comment (cahirwpz) [inline user/include/sys/mman.h]: Please use same values as mentioned in [mmap(2)](http://netbsd.gw.com/cgi-bin/man-cgi?mmap), ie. `MAP_ANONYMOUS`, `PROT_WRITE` and so on.

File: sys/sys_mmap.c [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    return -error;
+  return result;
+}
+
+vm_addr_t do_mmap(vm_addr_t addr, size_t length, vm_prot_t prot, int flags,
+                  int *error) {
+  thread_t *td = thread_self();
+  vm_map_t *vmap = td->td_uspace;
+
+  if (!vmap) {
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: This should be a fatal error. Translate to `assert` please.

**Comment by rafalcieslak** [comment]
All done.

I've also realized that with VFS interface in place, it will be pretty simple to implement support for non-anonymous (file-backed) mappings. All it would take is adding a new pager which keeps a reference to an open `file_t`.

File: sys/sys_mmap.c [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      /* No memory was found following the hint. Search again entire address
+         space. */
+      if (vm_map_findspace_nolock(vmap, MMAP_LOW_ADDR, length, &addr) != 0) {
+        /* Still no memory found. */
+        *error = ENOMEM;
+        return MMAP_FAILED;
+      }
+    }
+
+    /* Create new vm map entry for this allocation. Temporarily use permissive
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: I think this comment is outdated.

File: sys/sys_mmap.c [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  if (!flags & MMAP_FLAG_ANONYMOUS) {
+    log("Non-anonymous memory mappings are not yet implemented.");
+    *error = EINVAL;
+    return MMAP_FAILED;
+  }
+
+  {
+    rw_scoped_enter(&vmap->rwlock, RW_WRITER);
+
+    length = roundup(length, PAGESIZE);
```
Comment (cahirwpz) [inline sys/sys_mmap.c]: Only acquire lock for actions that require it. Please move lines 47-54 outside the scope.

File: sys/Makefile [hunk 10]
```diff
@@ -28,6 +28,7 @@ SOURCES_C  = \
 	startup.c \
 	sync.c \
 	sysent.c \
+	sys_mmap.c \
```
Comment (cahirwpz) [inline sys/Makefile]: If `sys_*.c` files are meant to contain syscall implementations I'd like this idea to be applied consistently throught code base.

**Comment by rafalcieslak** [comment]
The Travis test issue is a false positive, and I propose a fix for it in #250.

**Pull request merged**